src/ast/passes/codegen_llvm.h

```diff
@@ -35,51 +36,132 @@ struct VariableLLVM {
   llvm::Type *type;
 };
 
-class CodegenLLVM : public Visitor<CodegenLLVM> {
+// ScopedExpr ties an SSA value to a "delete" function, that typically will end
+// the lifetime of some needed storage. You must explicitly construct a
+// ScopedExpr, and can either construct from:
+// * A value only, with no associated function when out of scope.
+// * A value and associated function to run when out of scope.
+// * A value and another ScopedExpr, whose lifetime will be preserved until
+//   this value is out of scope.
+class ScopedExpr {
+public:
+  // Neither a value nor a deletion method.
+  explicit ScopedExpr() : value_(nullptr), deleter_(std::nullopt)
+  {
+  }
+
+  // Value only.
+  explicit ScopedExpr(Value *value) : value_(value), deleter_(std::nullopt)
+  {
+  }
+
+  // Value with an explicit deletion method.
+  explicit ScopedExpr(Value *value, llvm::unique_function<void()> &&deleter)
+      : value_(value), deleter_(std::move(deleter))
+  {
+  }
+
+  // Value with another ScopedExpr whose lifetime should be bound.
+  explicit ScopedExpr(Value *value, ScopedExpr &&other)
+      : value_(value), deleter_(std::move(other.deleter_))
+  {
+    other.deleter_.reset();
+  }
+
+  ScopedExpr(ScopedExpr &&other)
+      : value_(other.value_), deleter_(std::move(other.deleter_))
+  {
+    other.deleter_.reset();
+  }
+
+  ScopedExpr &operator=(ScopedExpr &&other)
+  {
+    value_ = other.value_;
+    deleter_.swap(other.deleter_);
+    return *this;
+  }
+
+  ScopedExpr(const ScopedExpr &other) = delete;
+  ScopedExpr &operator=(const ScopedExpr &other) = delete;
+
+  ~ScopedExpr()
+  {
+    if (deleter_) {
+      std::move(deleter_).value()();
+    }
+  }
+
+  Value *value()
+  {
+    return value_;
+  }
+
+  // Convenience method to override the current value, but preserve all the
+  // scope and deleter properties. Typically will be used for casts, etc.
+  void setValue(Value *value)
```

Comment:
This feels a bit weird to me. Maybe it should be something like `wrapValue` or another way to indicate that the underlying value doesn't need it's lifetime ended here. Because if you were truly replacing the value instead of just calling `CreatePtrToInt` then you'd need a different `deleter` right? Or am I missing something?